fix Issue #1409 - loadUnaligned produces incorrect inline LLVM IR

MartinNowak opened the pull request:
- apparently newer llvm versions use a different syntax

https://github.com/ldc-developers/ldc/issues/1409

MartinNowak commented:
I'm getting a warning that I can't interpret, maybe it's related to the change.

```
warning: Linking two modules of different data layouts: '<string>' is '' whereas '/home/dawg/Code/D/bug.d' is 'e-m:e-i64:64-f80:128-n8:16:32:64-S128'
```

dnadlinger commented:
The "inline IR" is lowered to a function in its own "mini" LLVM module (that is parsed from the CTFE'd string), which is then linked into the main module used for the rest of the codegen. Apparently, we fail to set the data layout for the mini-module. This should neither be directly related to the change nor affect the tests. It seems more like inline IR support has generally fallen a bit into disrepair (we didn't use to have a IR-level test suite when the feature was first implemented).

MartinNowak commented:
This change won't work w/ llvm 3.5.z, but I guess that's fine.

JohanEngelen commented:
@MartinNowak The minimum required LLVM version at the moment is 3.5.

JohanEngelen commented:
@MartinNowak You can use version(LDC_LLVM_30x) for that purpose. For example, one runnable inlineIR test we have reads:

``` D
// Syntax of load changed with LLVM 3.7
version(LDC_LLVM_305) version = PREDATES_307;
version(LDC_LLVM_306) version = PREDATES_307;
```

Looking at that file, I think you should use the same logic (it is the same syntax change of loads that you are facing)

redstar commented:
Just add `version(LDC_LLVM_305) { /* old stuff */ } else { ... }`.

MartinNowak committed:
fix Issue #1409 - loadUnaligned produces incorrect inline LLVM IR

- load syntax was changed w/ llvm 3.7.0
  http://llvm.org/releases/3.7.0/docs/LangRef.html#load-instruction

MartinNowak commented:
Updated, syntax was changed w/ 3.7.0.

JohanEngelen commented on src/ldc/simd.di:
```diff
@@ -204,14 +209,44 @@ if(is(typeof(llvmVecType!V)))
     alias inlineIR!(ir, V, T*) loadUnaligned;
 }
 
+unittest
```
unittesting is not done for *.di files. So we either have to enable that in the cmake script or move these tests elsewhere. In my opinion, it is much nicer to remove this testing "clutter" out of this file and into an IR test (possibly with an ASM test).

dnadlinger replied:
I'll take care of that after merge-2.071 is in master (hopefully in a couple of minutes).

JohanEngelen replied:
@klickverbot ping ;-)  There is also this: https://github.com/ldc-developers/ldc/issues/1518

dnadlinger replied:
Darn, I think this was at DConf, but the merge took longer than expected (until I pushed on accident^^). Please take over.

MartinNowak replied:
No idea how your testsuite works, adding a `test/ir/Makefile` prolly wouldn't work. Could you move this to an appropriate place?

JohanEngelen replied:
I'll try to take care of it @MartinNowak .
If you are interested, here is more information: https://wiki.dlang.org/LDC_Lit-based_testsuite

9il commented on src/ldc/simd.di:
```diff
@@ -204,14 +209,44 @@ if(is(typeof(llvmVecType!V)))
+            %r = load `~llvmV~`, `~llvmV~`* %p, align 1
+            ret `~llvmV~` %r`;
 
     alias inlineIR!(ir, V, T*) loadUnaligned;
```
Should this be `const(T)*`?

MartinNowak replied:
Maybe, wasn't part of the PR though.

9il commented on src/ldc/simd.di:
```diff
@@ -204,14 +209,44 @@ if(is(typeof(llvmVecType!V)))
 
+unittest
+{
+    float[4] f4buf;
```
please add tests for const and immutable data

9il commented:
See also #76

JohanEngelen commented:
@MartinNowak Can you remove the unittests from this PR? Thanks.

JohanEngelen commented:
@MartinNowak I have included this PR in this one: https://github.com/ldc-developers/ldc/pull/1526. So no work needed any more :-)
Thanks for the work.

MartinNowak commented:
Thanks
